sql: write all index entries for delete preserving indexes

stevendanna opened the pull request:
When updating a row, we calculate the index entries expected for a
given index using the old value and the new value of the row. To avoid
writing data unnecessarily, if an old and new index entry is the same,
we skip writing it.

This, however, is a problem for index additions. Consider the
following:

Assume we have a table with a schema such as

    CREATE TABLE t (
        pk INT PRIMARY KEY,
        a INT,
        b INT,
        c INT,
        FAMILY (pk, a),
        FAMILY (b),
        FAMILY (c)
    )

Then suppose we have the following series of events:

    t0: INSERT INTO t (pk, a, b, c) VALUES (1, 1, 1, 1)
    t1: CREATE INDEX t (a) STORING (b, c) // i2 added to table desc in BACKFILLING, i3 in DELETE_ONLY
    t2: i3 moves to DELETE_AND_WRITE_ONLY
    t3: AddSSTable based backfill of i2
    t4: i2 moves to DELETE_ONLY
    t5: UPDATE t SET b = 2 WHERE pk = 1;

Currently, the update at `t5` will produce the following KV
operations:

```
Put /Table/106/1/1/1/1 -> /INT/42               // Primary index entry for FAMILY (b)
Put /Table/106/3/1/1/1/1 -> /BYTES/0x0a030a3354 // Temporary index entry for FAMILY (b)
Del /Table/106/2/1/1/0
Del /Table/106/2/1/1/1/1
Del /Table/106/2/1/1/2/1
```

Note that we've deleted all entries from index 2 because it is in
DELETE_ONLY. And, in index 3, we only write out a value for the column
family that changed.  But, when we merge i3 into i2, the expected
index entries for the updated row will be incomplete as the expected
entry for the column family containing `c` will not be present.

To avoid this case, we write _all_ index entries for temporary
indexes. That is, in the new code, the update at t5 would produce the
following KV operations:

```
Put /Table/106/1/1/1/1 -> /INT/42               // Primary index entry for FAMILY (b)
Put /Table/106/3/1/1/0 -> /BYTES/0x0a0103
Put /Table/106/3/1/1/1/1 -> /BYTES/0x0a030a3354 // Temporary index entry for FAMILY (b)
Put /Table/106/3/1/1/2/1 -> /BYTES/0x0a030a4302 // Temporary index entry for FAMILY (c)
Del /Table/106/2/1/1/0
Del /Table/106/2/1/1/1/1
Del /Table/106/2/1/1/2/1
```

It's unfortunate that our index validation does not catch this case of
an invalid index creation.

Release justification: Critical correctness fix for create index
schema changes

Release note: None

stevendanna committed:
sql: write all index entries for delete preserving indexes

When updating a row, we calculate the index entries expected for a
given index using the old value and the new value of the row. To avoid
writing data unnecessarily, if an old and new index entry is the same,
we skip writing it.

This, however, is a problem for index additions. Consider the
following:

Assume we have a table with a schema such as

    CREATE TABLE t (
        pk INT PRIMARY KEY,
        a INT,
        b INT,
        c INT,
        FAMILY (pk, a),
        FAMILY (b),
        FAMILY (c)
    )

Then suppose we have the following series of events:

    t0: INSERT INTO t (pk, a, b, c) VALUES (1, 1, 1, 1)
    t1: CREATE INDEX t (a) STORING (b, c) // i2 added to table desc in BACKFILLING, i3 in DELETE_ONLY
    t2: i3 moves to DELETE_AND_WRITE_ONLY
    t3: AddSSTable based backfill of i2
    t4: i2 moves to DELETE_ONLY
    t5: UPDATE t SET b = 2 WHERE pk = 1;

Currently, the update at `t5` will produce the following KV
operations:

```
Put /Table/106/1/1/1/1 -> /INT/42               // Primary index entry for FAMILY (b)
Put /Table/106/3/1/1/1/1 -> /BYTES/0x0a030a3354 // Temporary index entry for FAMILY (b)
Del /Table/106/2/1/1/0
Del /Table/106/2/1/1/1/1
Del /Table/106/2/1/1/2/1
```

Note that we've deleted all entries from index 2 because it is in
DELETE_ONLY. And, in index 3, we only write out a value for the column
family that changed.  But, when we merge i3 into i2, the expected
index entries for the updated row will be incomplete as the expected
entry for the column family containing `c` will not be present.

To avoid this case, we write _all_ index entries for temporary
indexes. That is, in the new code, the update at t5 would produce the
following KV operations:

```
Put /Table/106/1/1/1/1 -> /INT/42               // Primary index entry for FAMILY (b)
Put /Table/106/3/1/1/0 -> /BYTES/0x0a0103
Put /Table/106/3/1/1/1/1 -> /BYTES/0x0a030a3354 // Temporary index entry for FAMILY (b)
Put /Table/106/3/1/1/2/1 -> /BYTES/0x0a030a4302 // Temporary index entry for FAMILY (c)
Del /Table/106/2/1/1/0
Del /Table/106/2/1/1/1/1
Del /Table/106/2/1/1/2/1
```

It's unfortunate that our index validation does not catch this case of
an invalid index creation.

Release justification: Critical correctness fix for create index
schema changes

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77464)
<!-- Reviewable:end -->

ajwerner commented:
I don't understand this example. In particular, why is there ever an entry for column `FAMILY (b)` in an index which is `INDEX (a)`? Is the index supposed to be `(a, b)` or something like that? It's very possible that I don't have a strong enough grasp on our key encoding.

stevendanna commented:
@ajwerner You are correct. I changed the example from an ALTER INDEX example to a CREATE INDEX example while writing the commit message and created an incomplete example.  I've pushed a new commit message with a more complete example and added the example as a test as well.

--

If we are worried about the extent to which this will increase the write load during index creation, I suppose a follow up could be to have different behaviour on the temporary index based on the state of the index it is tracking writes for.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

I wish so so badly that the `row` package had some nice data-driven tests. This is some serious testing-at-a-distance. 

> Reviewed 4 of 4 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77464#-:-Mxf8vqQAA9_tyDlUuMV:brz41cf)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rhu713)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> If we are worried about the extent to which this will increase the write load during index creation, I suppose a follow up could be to have different behaviour on the temporary index based on the state of the index it is tracking writes for.


I don't have concern about the increased write load. I have concern about increased contention. There is a rather major implication here which we shouldn't take lightly: during index creation, for indexes which involve multiple column families in their stored set, we won't get the benefits of column families to avoid contention. I think we should file an issue about it and move on. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77464#-:-Mxf98WV4AEA3G1TgXsO:b-t6o6cx)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rhu713)



<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=ajwerner

stevendanna commented:
I opened this here: https://github.com/cockroachdb/cockroach/issues/77530
